crates/build/src/docker.rs

```diff
@@ -0,0 +1,563 @@
+
+use std::{
+    collections::{
+        hash_map::DefaultHasher,
+        HashMap,
+    },
+    hash::{
+        Hash,
+        Hasher,
+    },
+    io::BufReader,
+    path::{
+        Path,
+        PathBuf,
+    },
+    time::Duration,
+};
+
+use anyhow::{
+    Context,
+    Result,
+};
+use bollard::{
+    container::{
+        Config,
+        CreateContainerOptions,
+        ListContainersOptions,
+        StopContainerOptions,
+        WaitContainerOptions,
+    },
+    errors::Error,
+    image::{
+        CreateImageOptions,
+        ListImagesOptions,
+    },
+    service::{
+        ContainerWaitResponse,
+        HostConfig,
+        ImageSummary,
+        Mount,
+        MountTypeEnum,
+    },
+    Docker,
+};
+use contract_metadata::ContractMetadata;
+use indicatif::{
+    ProgressBar,
+    ProgressStyle,
+};
+use tokio_stream::StreamExt;
+
+use crate::{
+    maybe_println,
+    BuildArtifacts,
+    BuildResult,
+    BuildSteps,
+    CrateMetadata,
+    ExecuteArgs,
+    Verbosity,
+};
+
+use colored::Colorize;
+
+const IMAGE: &str = "paritytech/contracts-verifiable";
+// We assume the docker image contains the same tag as the current version of the crate
+const VERSION: &str = env!("CARGO_PKG_VERSION");
+
+const MOUNT_DIR: &str = "/contract";
+
+#[derive(Clone, Debug, Default)]
+pub enum ImageVariant {
+    #[default]
+    Default,
+    Custom(String),
+}
+
+/// Launches the docker container to execute verifiable build
+pub fn docker_build(args: ExecuteArgs) -> Result<BuildResult> {
+    let ExecuteArgs {
+        manifest_path,
+        verbosity,
+        output_type,
+        target,
+        build_artifact,
+        image,
+        ..
+    } = args;
+    tokio::runtime::Builder::new_multi_thread()
+        .enable_all()
+        .build()?
+        .block_on(async {
+            let mut build_steps = BuildSteps::new();
+
+            build_steps.set_total_steps(3);
+            if build_artifact == BuildArtifacts::CodeOnly {
+                build_steps.set_total_steps(2);
+            }
+
+            let crate_metadata = CrateMetadata::collect(&manifest_path, target)?;
+            let mut manifest_dir_option = crate_metadata.manifest_path.directory();
+            let empty_path = PathBuf::new();
+            let contract_dir = manifest_dir_option.get_or_insert(&empty_path);
+            let host_dir = std::env::current_dir()?;
+            let build_result_path = contract_dir.join("target/build_result.json");
+            let args = compose_build_args()?;
+
+            let client = Docker::connect_with_socket_defaults().map_err(|e| {
+                anyhow::anyhow!("{}\nDo you have the docker engine installed in path?", e)
+            })?;
+            let _ = client.ping().await.map_err(|e| {
+                anyhow::anyhow!("{}\nIs your docker engine up and running?", e)
+            })?;
+            let build_image =
+                get_image(client.clone(), image, &verbosity, &mut build_steps).await?;
```

Comment:
I don't quite understand the purpose of this step of manually fetching the image. Using raw docker commands e.g. `docker run` or `docker create`, you can just supply the image name and docker will take care of pulling it to the local machine?

Reply:
* First, I am working with the docker API, so some of the operations need to be explicit. So, the API still requires us to "create" an image: https://docs.rs/bollard/latest/bollard/struct.Docker.html#method.create_image 
* Second, if we just specify the image upon creating the container, then when inspecting the container, it does not show the image name, only the digest, i.e.:
```
...
"Image": "sha256:022e29e85a19c24a555744d2c44fbc95a15d3b70df5bb5d6fc04c0c79b182698",
...
```
* Lastly, since we need to explicitly "create" an image. It allows us to show a progress bar for the user of the download. If the image is large, then we want to show the user how much time it takes. The idea behind verifiable builds is that we assume that the user does not have any experience with the docker and image/container management. Therefore, providing this information is just a nice thing to have.

Reply:
> First, I am working with the docker API, so some of the operations need to be explicit. So, the API still requires us to "create" an image: https://docs.rs/bollard/latest/bollard/struct.Docker.html#method.create_image

Ah okay so the `docker` client has the logic which you are replicating here. 

I did test without explicitly fetching the image but it turns out I still had a local image from a previous run. I was missing the `--all` from `docker image prune` to clear all images.

> Second, if we just specify the image upon creating the container, then when inspecting the container, it does not show the image name, only the digest, i.e.:

I was able to get this to work, but I guess it is moot now:

![image](https://github.com/paritytech/cargo-contract/assets/75586/51bbc89f-e9cf-49b0-926c-f081765848f0)